File: swim-system-java/swim-mesh-java/swim.runtime/src/main/java/swim/runtime/warp/WarpDownlinkModel.java
// Copyright 2015-2019 SWIM.AI inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package swim.runtime.warp;

import swim.runtime.DownlinkRelay;
import swim.runtime.DownlinkView;
import swim.runtime.Push;
import swim.structure.Value;
import swim.uri.Uri;
import swim.warp.CommandMessage;
import swim.warp.EventMessage;
import swim.warp.LinkRequest;
import swim.warp.LinkedResponse;
import swim.warp.SyncRequest;
import swim.warp.SyncedResponse;
import swim.warp.UnlinkRequest;
import swim.warp.UnlinkedResponse;

public abstract class WarpDownlinkModel<View extends WarpDownlinkView> extends WarpDownlinkModem<View> {
  public WarpDownlinkModel(Uri meshUri, Uri hostUri, Uri nodeUri, Uri laneUri,
                           float prio, float rate, Value body) {
    super(meshUri, hostUri, nodeUri, laneUri, prio, rate, body);
  }

  @Override
  public final boolean keepLinked() {
    final Object views = this.views;
    if (views instanceof WarpDownlinkView) {
      return ((WarpDownlinkView) views).keepLinked();
    } else if (views instanceof DownlinkView[]) {
      final DownlinkView[] viewArray = (DownlinkView[]) views;
      for (int i = 0, n = viewArray.length; i < n; i += 1) {
        if (((WarpDownlinkView) viewArray[i]).keepLinked()) {
          return true;
        }
      }
    }
    return false;
  }

  @Override
  public final boolean keepSynced() {
    final Object views = this.views;
    if (views instanceof WarpDownlinkView) {
      return ((WarpDownlinkView) views).keepSynced();
    } else if (views instanceof DownlinkView[]) {
      final DownlinkView[] viewArray = (DownlinkView[]) views;
      for (int i = 0, n = viewArray.length; i < n; i += 1) {
        if (((WarpDownlinkView) viewArray[i]).keepSynced()) {
          return true;
        }
      }
    }
    return false;
  }

  @Override
  protected void pushDownEvent(Push<EventMessage> push) {
    onEvent(push.message());
    new WarpDownlinkRelayOnEvent<View>(this, push).run();
  }

  @Override
  protected void pushDownLinked(Push<LinkedResponse> push) {
    didLink(push.message());
    new WarpDownlinkRelayDidLink<View>(this, push).run();
  }

  @Override
  protected void pushDownSynced(Push<SyncedResponse> push) {
    didSync(push.message());
    new WarpDownlinkRelayDidSync<View>(this, push).run();
  }

  @Override
  protected void pushDownUnlinked(Push<UnlinkedResponse> push) {
    didUnlink(push.message());
    new WarpDownlinkRelayDidUnlink<View>(this, push).run();
  }

  @Override
  protected void pullUpCommand(CommandMessage message) {
    onCommand(message);
    new WarpDownlinkRelayWillCommand<View>(this, message).run();
  }

  @Override
  protected void pullUpLink(LinkRequest request) {
    willLink(request);
    new WarpDownlinkRelayWillLink<View>(this, request).run();
  }

  @Override
  protected void pullUpSync(SyncRequest request) {
    willSync(request);
    new WarpDownlinkRelayWillSync<View>(this, request).run();
  }

  @Override
  protected void pullUpUnlink(UnlinkRequest request) {
    willUnlink(request);
    new WarpDownlinkRelayWillUnlink<View>(this, request).run();
  }
}

final class WarpDownlinkRelayOnEvent<View extends WarpDownlinkView> extends DownlinkRelay<WarpDownlinkModel<View>, View> {
  final Push<EventMessage> push;

  WarpDownlinkRelayOnEvent(WarpDownlinkModel<View> model, Push<EventMessage> push) {
    super(model, 2);
    this.push = push;
  }

  @Override
  protected boolean runPhase(View view, int phase, boolean preemptive) {
    if (phase == 0) {
      final EventMessage message = this.push.message();
      if (preemptive) {
        view.downlinkWillReceive(message);
      }
      return view.dispatchWillReceive(message.body(), preemptive);
    } else if (phase == 1) {
      final EventMessage message = this.push.message();
      if (preemptive) {
        view.downlinkDidReceive(message);
      }
      return view.dispatchDidReceive(message.body(), preemptive);
    } else {
      throw new AssertionError(); // unreachable
    }
  }

  @Override
  protected void done() {
    this.push.bind();
    this.model.cueDown();
  }
}

final class WarpDownlinkRelayWillCommand<View extends WarpDownlinkView> extends DownlinkRelay<WarpDownlinkModel<View>, View> {
  final CommandMessage message;

  WarpDownlinkRelayWillCommand(WarpDownlinkModel<View> model, CommandMessage message) {
    super(model);
    this.message = message;
  }

  @Override
  protected boolean runPhase(View view, int phase, boolean preemptive) {
    if (phase == 0) {
      if (preemptive) {
        view.downlinkWillCommand(this.message);
      }
      return view.dispatchWillCommand(this.message.body(), preemptive);
    } else {
      throw new AssertionError(); // unreachable
    }
  }
}

final class WarpDownlinkRelayWillLink<View extends WarpDownlinkView> extends DownlinkRelay<WarpDownlinkModel<View>, View> {
  final LinkRequest request;

  WarpDownlinkRelayWillLink(WarpDownlinkModel<View> model, LinkRequest request) {
    super(model);
    this.request = request;
  }

  @Override
  protected boolean runPhase(View view, int phase, boolean preemptive) {
    if (phase == 0) {
      if (preemptive) {
        view.downlinkWillLink(this.request);
      }
      return view.dispatchWillLink(preemptive);
    } else {
      throw new AssertionError(); // unreachable
    }
  }
}

final class WarpDownlinkRelayDidLink<View extends WarpDownlinkView> extends DownlinkRelay<WarpDownlinkModel<View>, View> {
  final Push<LinkedResponse> push;

  WarpDownlinkRelayDidLink(WarpDownlinkModel<View> model, Push<LinkedResponse> push) {
    super(model);
    this.push = push;
  }

  @Override
  protected boolean runPhase(View view, int phase, boolean preemptive) {
    if (phase == 0) {
      if (preemptive) {
        view.downlinkDidLink(this.push.message());
      }
      return view.dispatchDidLink(preemptive);
    } else {
      throw new AssertionError(); // unreachable
    }
  }

  @Override
  protected void done() {
    this.push.bind();
    this.model.cueDown();
  }
}

final class WarpDownlinkRelayWillSync<View extends WarpDownlinkView> extends DownlinkRelay<WarpDownlinkModel<View>, View> {
  final SyncRequest request;

  WarpDownlinkRelayWillSync(WarpDownlinkModel<View> model, SyncRequest request) {
    super(model);
    this.request = request;
  }

  @Override
  protected boolean runPhase(View view, int phase, boolean preemptive) {
    if (phase == 0) {
      if (preemptive) {
        view.downlinkWillSync(this.request);
      }
      return view.dispatchWillSync(preemptive);
    } else {
      throw new AssertionError(); // unreachable
    }
  }
}

final class WarpDownlinkRelayDidSync<View extends WarpDownlinkView> extends DownlinkRelay<WarpDownlinkModel<View>, View> {
  final Push<SyncedResponse> push;

  WarpDownlinkRelayDidSync(WarpDownlinkModel<View> model, Push<SyncedResponse> push) {
    super(model);
    this.push = push;
  }

  @Override
  protected boolean runPhase(View view, int phase, boolean preemptive) {
    if (phase == 0) {
      if (preemptive) {
        view.downlinkDidSync(this.push.message());
      }
      return view.dispatchDidSync(preemptive);
    } else {
      throw new AssertionError(); // unreachable
    }
  }

  @Override
  protected void done() {
    this.push.bind();
    this.model.cueDown();
  }
}

final class WarpDownlinkRelayWillUnlink<View extends WarpDownlinkView> extends DownlinkRelay<WarpDownlinkModel<View>, View> {
  final UnlinkRequest request;

  WarpDownlinkRelayWillUnlink(WarpDownlinkModel<View> model, UnlinkRequest request) {
    super(model);
    this.request = request;
  }

  @Override
  protected boolean runPhase(View view, int phase, boolean preemptive) {
    if (phase == 0) {
      if (preemptive) {
        view.downlinkWillUnlink(this.request);
      }
      return view.dispatchWillUnlink(preemptive);
    } else {
      throw new AssertionError(); // unreachable
    }
  }
}

final class WarpDownlinkRelayDidUnlink<View extends WarpDownlinkView> extends DownlinkRelay<WarpDownlinkModel<View>, View> {
  final Push<UnlinkedResponse> push;

  WarpDownlinkRelayDidUnlink(WarpDownlinkModel<View> model, Push<UnlinkedResponse> push) {
    super(model);
    this.push = push;
  }

  @Override
  protected boolean runPhase(View view, int phase, boolean preemptive) {
    if (phase == 0) {
      if (preemptive) {
        view.downlinkDidUnlink(this.push.message());
      }
      return view.dispatchDidUnlink(preemptive);
    } else {
      throw new AssertionError(); // unreachable
    }
  }

  @Override
  protected void done() {
    this.push.bind();
    // Don't cueDown model after unlinked.
  }
}

File: swim-js/swim-host/swim-client/main/WarpDownlinkModel.ts
// Copyright 2015-2023 Nstream, inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import type {Mutable} from "@swim/util";
import {Property} from "@swim/component";
import type {ComponentFlags} from "@swim/component";
import {Component} from "@swim/component";
import type {Uri} from "@swim/uri";
import type {ValueLike} from "@swim/structure";
import {Value} from "@swim/structure";
import type {EventMessage} from "@swim/warp";
import {LinkRequest} from "@swim/warp";
import type {LinkedResponse} from "@swim/warp";
import {SyncRequest} from "@swim/warp";
import type {SyncedResponse} from "@swim/warp";
import type {UnlinkRequest} from "@swim/warp";
import type {UnlinkedResponse} from "@swim/warp";
import type {WarpDownlink} from "./WarpDownlink";
import {WarpHost} from "./"; // forward import

/** @internal */
export class WarpDownlinkModel extends Component {
  constructor(hostUri: Uri, nodeUri: Uri, laneUri: Uri,
              prio: number, rate: number, body: Value) {
    super();
    this.hostUri = hostUri;
    this.nodeUri = nodeUri;
    this.laneUri = laneUri;
    this.prio = prio;
    this.rate = rate;
    this.body = body;
    this.views = null;
  }

  readonly hostUri: Uri;

  readonly nodeUri: Uri;

  readonly laneUri: Uri;

  readonly prio: number;

  readonly rate: number;

  readonly body: Value;

  readonly views: ReadonlySet<WarpDownlink> | null;

  keepLinked(): boolean {
    const views = this.views;
    if (views !== null) {
      for (const view of views) {
        if (view.relinks) {
          return true;
        }
      }
    }
    return false;
  }

  keepSynced(): boolean {
    const views = this.views;
    if (views !== null) {
      for (const view of views) {
        if (view.syncs) {
          return true;
        }
      }
    }
    return false;
  }

  @Property({valueType: Boolean, value: true, inherits: true})
  readonly online!: Property<this, boolean>;

  get connected(): boolean {
    return (this.flags & WarpDownlinkModel.ConnectedFlag) !== 0;
  }

  /** @internal */
  setConnected(connected: boolean): void {
    if (connected && (this.flags & WarpDownlinkModel.ConnectedFlag) === 0) {
      this.setFlags(this.flags | WarpDownlinkModel.ConnectedFlag);
      this.willConnect();
      this.onConnect();
      this.didConnect();
    } else if (!connected && (this.flags & WarpDownlinkModel.ConnectedFlag) !== 0) {
      this.setFlags(this.flags & ~WarpDownlinkModel.ConnectedFlag);
      this.willDisconnect();
      this.onDisconnect();
      this.didDisconnect();
    }
  }

  protected willConnect(): void {
    // hook
  }

  protected onConnect(): void {
    const views = this.views;
    if (views !== null) {
      for (const view of views) {
        view.hostDidConnect();
      }
    }
  }

  protected didConnect(): void {
    if (this.keepSynced()) {
      this.sync();
    } else {
      this.link();
    }
  }

  protected willDisconnect(): void {
    this.setFlags(this.flags & ~WarpDownlinkModel.DownlinkMask);
  }

  protected onDisconnect(): void {
    const views = this.views;
    if (views !== null) {
      for (const view of views) {
        view.hostDidDisconnect();
      }
    }
  }

  protected didDisconnect(): void {
    if (!this.keepLinked()) {
      this.close();
    }
  }

  get linked(): boolean {
    return (this.flags & WarpDownlinkModel.LinkedFlag) !== 0;
  }

  get synced(): boolean {
    return (this.flags & WarpDownlinkModel.SyncedFlag) !== 0;
  }

  get authenticated(): boolean {
    return (this.flags & WarpHost.AuthenticatedFlag) !== 0;
  }

  /** @internal */
  setAuthenticated(authenticated: boolean): void {
    if (authenticated === ((this.flags & WarpHost.AuthenticatedFlag) !== 0)) {
      return;
    }
    this.willSetAuthenticated(authenticated);
    if (authenticated) {
      this.setFlags(this.flags | WarpHost.AuthenticatedFlag);
    } else {
      this.setFlags(this.flags & ~WarpHost.AuthenticatedFlag);
    }
    this.onSetAuthenticated(authenticated);
    this.didSetAuthenticated(authenticated);
  }

  protected willSetAuthenticated(authenticated: boolean): void {
    // hook
  }

  protected onSetAuthenticated(authenticated: boolean): void {
    // hook
  }

  protected didSetAuthenticated(authenticated: boolean): void {
    // hook
  }

  get deauthenticated(): boolean {
    return (this.flags & WarpHost.AuthenticatedFlag) !== 0;
  }

  /** @internal */
  setDeauthenticated(deauthenticated: boolean): void {
    if (deauthenticated === ((this.flags & WarpHost.DeauthenticatedFlag) !== 0)) {
      return;
    }
    this.willSetDeauthenticated(deauthenticated);
    if (deauthenticated) {
      this.setFlags(this.flags | WarpHost.DeauthenticatedFlag);
    } else {
      this.setFlags(this.flags & ~WarpHost.DeauthenticatedFlag);
    }
    this.onSetDeauthenticated(deauthenticated);
    this.didSetDeauthenticated(deauthenticated);
  }

  protected willSetDeauthenticated(deauthenticated: boolean): void {
    // hook
  }

  protected onSetDeauthenticated(deauthenticated: boolean): void {
    // hook
  }

  protected didSetDeauthenticated(deauthenticated: boolean): void {
    // hook
  }

  @Property({valueType: Value, value: Value.absent(), inherits: true})
  readonly session!: Property<this, Value>;

  addDownlink(view: WarpDownlink): void {
    let views = this.views as Set<WarpDownlink> | null;
    if (views === null) {
      views = new Set<WarpDownlink>();
      (this as Mutable<this>).views = views;
    } else if (views.has(view)) {
      return;
    }
    views.add(view);
  }

  removeDownlink(view: WarpDownlink): void {
    const views = this.views as Set<WarpDownlink> | null;
    if (views === null || !views.has(view)) {
      return;
    }
    views.delete(view);
    if (views.size === 0) {
      const unlinkDelay = this.unlinkDelay.value;
      if (unlinkDelay < 0) {
        this.unlink();
      } else {
        setTimeout(this.doUnlink.bind(this), unlinkDelay);
      }
    }
    view.close();
  }

  sync(): void {
    const host = this.getAncestor(WarpHost);
    if (host === null) {
      return;
    }
    const nodeUri = host.unresolve(this.nodeUri);
    const request = new SyncRequest(nodeUri, this.laneUri, this.prio, this.rate, this.body);
    this.onSyncRequest(request);
    host.push(request);
  }

  link(): void {
    const host = this.getAncestor(WarpHost);
    if (host === null) {
      return;
    }
    const nodeUri = host.unresolve(this.nodeUri);
    const request = new LinkRequest(nodeUri, this.laneUri, this.prio, this.rate, this.body);
    this.onLinkRequest(request);
    host.push(request);
  }

  unlink(): void {
    this.setFlags(this.flags & ~WarpDownlinkModel.DownlinkMask | WarpDownlinkModel.UnlinkingFlag);
    const host = this.getAncestor(WarpHost);
    if (host === null) {
      return;
    }
    host.unlinkDownlink(this);
  }

  protected doUnlink(): void {
    if (this.views !== null && this.views.size === 0) {
      this.unlink();
    }
  }

  @Property({valueType: Number, value: 0, inherits: true})
  readonly unlinkDelay!: Property<this, number>;

  command(body: ValueLike): void {
    const host = this.getAncestor(WarpHost);
    if (host === null) {
      return;
    }
    body = Value.fromLike(body);
    this.onCommandMessage(body);
    host.command(this.nodeUri, this.laneUri, body);
  }

  onEventMessage(message: EventMessage, host: WarpHost): void {
    const views = this.views;
    if (views !== null) {
      for (const view of views) {
        view.onEventMessage(message);
      }
    }
  }

  onCommandMessage(body: Value): void {
    const views = this.views;
    if (views !== null) {
      for (const view of views) {
        view.onCommandMessage(body);
      }
    }
  }

  onLinkRequest(request: LinkRequest): void {
    this.setFlags(this.flags | WarpDownlinkModel.LinkingFlag);
    const views = this.views;
    if (views !== null) {
      for (const view of views) {
        view.onLinkRequest(request);
      }
    }
  }

  onLinkedResponse(response: LinkedResponse, host: WarpHost): void {
    this.setFlags(this.flags & ~WarpDownlinkModel.LinkingFlag | WarpDownlinkModel.LinkedFlag);
    const views = this.views;
    if (views !== null) {
      for (const view of views) {
        view.onLinkedResponse(response);
      }
    }
  }

  onSyncRequest(request: SyncRequest): void {
    this.setFlags(this.flags | WarpDownlinkModel.SyncingFlag);
    const views = this.views;
    if (views !== null) {
      for (const view of views) {
        view.onSyncRequest(request);
      }
    }
  }

  onSyncedResponse(response: SyncedResponse, host: WarpHost): void {
    this.setFlags(this.flags & ~WarpDownlinkModel.SyncingFlag | WarpDownlinkModel.SyncedFlag);
    const views = this.views;
    if (views !== null) {
      for (const view of views) {
        view.onSyncedResponse(response);
      }
    }
  }

  onUnlinkRequest(request: UnlinkRequest, host: WarpHost): void {
    this.setFlags(this.flags & ~(WarpDownlinkModel.LinkingFlag | WarpDownlinkModel.SyncingFlag) | WarpDownlinkModel.UnlinkingFlag);
    const views = this.views;
    if (views !== null) {
      for (const view of views) {
        view.onUnlinkRequest(request);
      }
    }
  }

  onUnlinkedResponse(response: UnlinkedResponse, host: WarpHost): void {
    this.setFlags(this.flags & ~WarpDownlinkModel.UnlinkingFlag);
    const views = this.views;
    if (views === null || views.size === 0 || (this.flags & WarpDownlinkModel.DownlinkMask) !== 0) {
      if (views !== null) {
        for (const view of views) {
          view.onUnlinkedResponse(response);
        }
      }
      this.close();
      return;
    }
    // Concurrently relinked.
    if (this.keepSynced()) {
      this.sync();
    } else {
      this.link();
    }
  }

  hostDidFail(error: unknown, host: WarpHost): void {
    const views = this.views;
    if (views !== null) {
      for (const view of views) {
        view.hostDidFail(error);
      }
    }
  }

  close(): void {
    this.remove();
  }

  protected override onUnmount(): void {
    super.onUnmount();
    const views = this.views;
    if (views !== null) {
      (this as Mutable<this>).views = null;
      for (const view of views) {
        view.close();
      }
    }
  }

  /** @internal */
  static readonly ConnectedFlag: ComponentFlags = 1 << (Component.FlagShift + 0);
  /** @internal */
  static readonly AuthenticatedFlag: ComponentFlags = 1 << (Component.FlagShift + 1);
  /** @internal */
  static readonly DeauthenticatedFlag: ComponentFlags = 1 << (Component.FlagShift + 2);
  /** @internal */
  static readonly LinkingFlag: ComponentFlags = 1 << (Component.FlagShift + 3);
  /** @internal */
  static readonly LinkedFlag: ComponentFlags = 1 << (Component.FlagShift + 4);
  /** @internal */
  static readonly SyncingFlag: ComponentFlags = 1 << (Component.FlagShift + 5);
  /** @internal */
  static readonly SyncedFlag: ComponentFlags = 1 << (Component.FlagShift + 6);
  /** @internal */
  static readonly UnlinkingFlag: ComponentFlags = 1 << (Component.FlagShift + 7);

  /** @internal */
  static readonly DownlinkMask: ComponentFlags = this.LinkingFlag
                                               | this.LinkedFlag
                                               | this.SyncingFlag
                                               | this.SyncedFlag
                                               | this.UnlinkingFlag;

  /** @internal */
  static override readonly FlagShift: number = Component.FlagShift + 8;
  /** @internal */
  static override readonly FlagMask: ComponentFlags = (1 << this.FlagShift) - 1;
}
